proxy/coco/src/peer/run_state.rs

```diff
@@ -0,0 +1,361 @@
+    /// Local peer is listening on a socket, but no connected peers yet.
+    Started(Instant),
+    /// The loccal peer lost connections to all peers.
+    Offline(Instant),
+    /// Phase where the local peer tries get up-to-date.
+    Syncing(Instant, usize),
+    /// Normal operation with other other peers being connected.
+    Online(Instant),
+}
+
+/// Set of knobs to change the behaviour of the [`RunState`].
+#[derive(Default)]
+pub struct Config {
+    /// Enables the syncing stage when coming online.
+    pub sync_on_startup: bool,
+}
+
+/// State kept for a running local peer.
+pub struct RunState {
+    /// Confiugration to change how input [`Event`]s are interpreted.
+    config: Config,
+    /// Tracking remote peers that have an active connection.
+    connected_peers: HashSet<PeerId>,
+    /// Current internal status.
+    status: Status,
+}
+
+impl From<Config> for RunState {
+    fn from(config: Config) -> Self {
+        Self {
+            config,
+            connected_peers: HashSet::new(),
+            status: Status::Stopped(Instant::now()),
+        }
+    }
+}
+
+impl RunState {
+    /// Constructs a new state.
+    #[cfg(test)]
+    const fn new(config: Config, connected_peers: HashSet<PeerId>, status: Status) -> Self {
+        Self {
+            config,
+            connected_peers,
+            status,
+        }
+    }
+
+    /// Applies the `input` and based on the current state transforms to the new state and in some
+    /// cases produes commands which should be executed in the appropriate sub-routines.
+    pub fn transition(&mut self, event: Event) -> Vec<Command> {
+        match (&self.status, event) {
+            // Go from [`Input::Stopped`] to [`Input::Offline`] once we are listening.
+            (Status::Stopped(_since), Event::Protocol(ProtocolEvent::Listening(_addr))) => {
+                self.status = Status::Started(Instant::now());
+
+                vec![]
+            },
+            // Sync with first incoming peer.
+            //
+            // In case the peer is configured to sync on startup we start syncing, otherwise we go
+            // online straight away.
+            (Status::Started(_since), Event::Protocol(ProtocolEvent::Connected(ref peer_id))) => {
+                self.connected_peers.insert(peer_id.clone());
+
+                if self.config.sync_on_startup {
+                    self.status = Status::Syncing(Instant::now(), 1);
+
+                    vec![
+                        Command::SyncPeer(peer_id.clone()),
+                        Command::StartSyncTimeout,
+                    ]
+                } else {
+                    self.status = Status::Online(Instant::now());
+
+                    vec![]
+                }
+            },
+            // Sync until configured maximum of peers is reached.
+            (Status::Syncing(since, syncs), Event::Protocol(ProtocolEvent::Connected(peer_id)))
+                if *syncs < SYNC_MAX_PEERS =>
```

Comment:
`SYNC_MAX_PEERS` definitely could be defined in `Config`, I would say

Reply:
For sure.